src/main/java/teammates/storage/api/DeadlineExtensionsDb.java

```diff
@@ -0,0 +1,227 @@
+            return;
+        }
+
+        List<DeadlineExtension> entitiesToUpdate = load().project()
+                .filter("courseId =", courseId)
+                .filter("userEmail =", oldEmail)
+                .filter("isInstructor =", isInstructor)
+                .list();
+
+        for (DeadlineExtension deadlineExtension : entitiesToUpdate) {
+            try {
+                updateDeadlineExtension(DeadlineExtensionAttributes
+                        .updateOptionsBuilder(courseId, deadlineExtension.getFeedbackSessionName(), oldEmail, isInstructor)
+                        .withNewEmail(newEmail)
+                        .build());
+            } catch (EntityDoesNotExistException ednee) {
+                // ignore: deadline extension entity already deleted
+            }
+        }
+    }
+
+    private DeadlineExtension getDeadlineExtensionEntity(String id) {
+        return load().id(id).now();
+    }
+
+    /**
+     * Deletes a deadline extension.
+     */
+    public void deleteDeadlineExtension(
+            String courseId, String feedbackSessionName, String userEmail, boolean isInstructor) {
+        assert courseId != null;
+        assert feedbackSessionName != null;
+        assert userEmail != null;
+
+        deleteEntity(Key.create(DeadlineExtension.class,
+                DeadlineExtension.generateId(courseId, feedbackSessionName, userEmail, isInstructor)));
+    }
+
+    /**
+     * Deletes deadline extensions using {@link AttributesDeletionQuery}.
+     */
+    public void deleteDeadlineExtensions(AttributesDeletionQuery query) {
+        assert query != null;
+        assert verifyValidDeletionQuery(query);
+
+        Query<DeadlineExtension> entitiesToDelete = load().project().filter("courseId =", query.getCourseId());
+
+        if (query.isFeedbackSessionNamePresent()) {
+            entitiesToDelete = entitiesToDelete.filter("feedbackSessionName =", query.getFeedbackSessionName());
+        } else if (query.isUserEmailPresent() && query.isIsInstructorPresent()) {
+            entitiesToDelete = entitiesToDelete.filter("userEmail =", query.getUserEmail());
+            entitiesToDelete = entitiesToDelete.filter("isInstructor =", query.getIsInstructor());
+        }
+
+        deleteEntity(entitiesToDelete.keys().list());
+    }
+
+    private boolean verifyValidDeletionQuery(AttributesDeletionQuery query) {
+        if (!query.isCourseIdPresent()) {
+            return false;
+        }
+
+        // Valid combination of parameters:
+        // courseId
+        // courseId, feedbackSessionName
+        // courseId, isInstructor, userEmail
+        return !query.isIsInstructorPresent() && !query.isUserEmailPresent() && !query.isFeedbackSessionNamePresent()
+                || !query.isIsInstructorPresent() && !query.isUserEmailPresent() && query.isFeedbackSessionNamePresent()
+                || query.isIsInstructorPresent() && query.isUserEmailPresent() && !query.isFeedbackSessionNamePresent();
```

Comment:
Maybe we should use variables here instead of comments, to make it self-documenting?
```suggestion
        boolean isValidForDeletingAllInCourse = !query.isIsInstructorPresent() && !query.isUserEmailPresent() && !query.isFeedbackSessionNamePresent();
        boolean isValidForDeletingAllInSession = !query.isIsInstructorPresent() && !query.isUserEmailPresent() && query.isFeedbackSessionNamePresent();
        boolean isValidForDeletingAllForUser = query.isIsInstructorPresent() && query.isUserEmailPresent() && !query.isFeedbackSessionNamePresent();
        return isValidForDeletingAllInCourse || isValidForDeletingAllInSession || isValidForDeletingAllForUser;
```

Reply:
Sure, I'll update now.